source/extensions/filters/http/json_to_metadata/filter.cc

```diff
@@ -320,41 +365,76 @@ void Filter::processBody(const Buffer::Instance* body, const Rules& rules,
     if (!result.ok()) {
       ENVOY_LOG(warn, fmt::format("{} key: {}", result.message(), keys.back()));
-      handleOnMissing(rule, struct_map);
+      handleOnMissing(rule, struct_map, filter_callback);
     }
   }
   success.inc();
 
-  finalizeDynamicMetadata(*decoder_callbacks_, struct_map, processing_finished_flag);
+  finalizeDynamicMetadata(filter_callback, struct_map, processing_finished_flag);
 }
 
 void Filter::processRequestBody() {
   processBody(decoder_callbacks_->decodingBuffer(), config_->requestRules(),
               request_processing_finished_, config_->stats().rq_success_,
-              config_->stats().rq_no_body_, config_->stats().rq_invalid_json_body_);
+              config_->stats().rq_no_body_, config_->stats().rq_invalid_json_body_,
+              *decoder_callbacks_);
+}
+
+void Filter::processResponseBody() {
+  processBody(encoder_callbacks_->encodingBuffer(), config_->responseRules(),
+              response_processing_finished_, config_->stats().rsp_success_,
+              config_->stats().rsp_no_body_, config_->stats().rsp_invalid_json_body_,
+              *encoder_callbacks_);
 }
 
 Http::FilterHeadersStatus Filter::decodeHeaders(Http::RequestHeaderMap& headers, bool end_stream) {
-  ASSERT(config_->doRequest());
+  ASSERT(config_->doRequest() || config_->doResponse());
+  if (!config_->doRequest()) {
+    return Http::FilterHeadersStatus::Continue;
+  }
   if (!config_->requestContentTypeAllowed(headers.getContentTypeValue())) {
     request_processing_finished_ = true;
     config_->stats().rq_mismatched_content_type_.inc();
     return Http::FilterHeadersStatus::Continue;
   }
 
   if (end_stream) {
-    handleAllOnMissing(config_->requestRules(), request_processing_finished_);
+    handleAllOnMissing(config_->requestRules(), request_processing_finished_, *decoder_callbacks_);
     config_->stats().rq_no_body_.inc();
     return Http::FilterHeadersStatus::Continue;
   }
   return Http::FilterHeadersStatus::StopIteration;
 }
 
+Http::FilterHeadersStatus Filter::encodeHeaders(Http::ResponseHeaderMap& headers, bool end_stream) {
+  ASSERT(config_->doRequest() || config_->doResponse());
+  if (!config_->doResponse()) {
+    return Http::FilterHeadersStatus::Continue;
+  }
+  if (!config_->responseContentTypeAllowed(headers.getContentTypeValue())) {
+    response_processing_finished_ = true;
+    config_->stats().rsp_mismatched_content_type_.inc();
+    return Http::FilterHeadersStatus::Continue;
+  }
+
+  if (end_stream) {
+    handleAllOnMissing(config_->responseRules(), response_processing_finished_,
+                       *encoder_callbacks_);
+    config_->stats().rsp_no_body_.inc();
+    return Http::FilterHeadersStatus::Continue;
+  }
+  return Http::FilterHeadersStatus::StopIteration;
+}
+
 Http::FilterDataStatus Filter::decodeData(Buffer::Instance& data, bool end_stream) {
-  ASSERT(config_->doRequest());
+  ASSERT(config_->doRequest() || config_->doResponse());
```

Comment:
remove the assert here and there.

Reply:
> remove the assert here and there.

I thought we want to check at least one of request/response rule is configured.

Reply:
We should check in control plane instead of here if we haven't

Reply:
Looks like we haven't checked. In FilterConfig constructor, throw if no rule is set. Check http header to metadata filter for reference.